### PR TITLE
fix: fix signatures of fftw{l,,f}_plan_guru_r2r

### LINKED ISSUE #378: Build failure due to incompatible-pointer-types: `const fftw_r2r_kind *` vs `int *` for `fftw_plan_guru_r2r`
Hi. I tried to build #370 by myself but failed on EndeavourOS, where

- GCC 14.1.1
- python 3.12.3

The messages are as follows:

```ShellSession
$ pip install .
…
      Compiling pyfftw/pyfftw.pyx because it depends on /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/Cython/Includes/libc/string.pxd.
      [1/1] Cythonizing pyfftw/pyfftw.pyx
      INFO:root:building 'pyfftw.pyfftw' extension
      INFO:root:gcc -fno-strict-overflow -DNDEBUG -g -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -ffat-lto-objects -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -fPIC -DPYFFTW_HAVE_DOUBLE=1 -DPYFFTW_HAVE_DOUBLE_OMP=1 -DPYFFTW_HAVE_DOUBLE_THREADS=0 -DPYFFTW_HAVE_DOUBLE_MULTITHREADING=1 -DPYFFTW_HAVE_DOUBLE_MPI=0 -DPYFFTW_HAVE_SINGLE=1 -DPYFFTW_HAVE_SINGLE_OMP=1 -DPYFFTW_HAVE_SINGLE_THREADS=0 -DPYFFTW_HAVE_SINGLE_MULTITHREADING=1 -DPYFFTW_HAVE_SINGLE_MPI=0 -DPYFFTW_HAVE_LONG=1 -DPYFFTW_HAVE_LONG_OMP=1 -DPYFFTW_HAVE_LONG_THREADS=0 -DPYFFTW_HAVE_LONG_MULTITHREADING=1 -DPYFFTW_HAVE_LONG_MPI=0 -DPYFFTW_HAVE_MPI=0 -I./venv/include -I/usr/include/python3.12 -I./include -I./pyfftw -I/tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include -I./venv/include -I/usr/include -I./venv/include -I/usr/include/python3.12 -c pyfftw/pyfftw.c -o build/temp.linux-x86_64-cpython-312/pyfftw/pyfftw.o
      In file included from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/ndarraytypes.h:1929,
                       from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                       from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/arrayobject.h:5,
                       from pyfftw/pyfftw.c:1306:
      /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
         17 | #warning "Using deprecated NumPy API, disable it with " \
            |  ^~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftw_plan_guru_r2r’:
      pyfftw/pyfftw.c:13236:164: error: passing argument 7 of ‘fftw_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13236 |   __pyx_r = ((void *)fftw_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((double *)__pyx_v__in), ((double *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                    ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                    |
            |                                                                                                                                                                    int *
      In file included from pyfftw/pyfftw.c:1313:
      /usr/include/fftw3.h:465:1: note: expected ‘const fftw_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        465 | FFTW_DEFINE_API(FFTW_MANGLE_DOUBLE, double, fftw_complex)
            | ^~~~~~~~~~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftwf_plan_guru_r2r’:
      pyfftw/pyfftw.c:13372:163: error: passing argument 7 of ‘fftwf_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13372 |   __pyx_r = ((void *)fftwf_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((float *)__pyx_v__in), ((float *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                   ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                   |
            |                                                                                                                                                                   int *
      /usr/include/fftw3.h:466:1: note: expected ‘const fftwf_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        466 | FFTW_DEFINE_API(FFTW_MANGLE_FLOAT, float, fftwf_complex)
            | ^~~~~~~~~~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftwl_plan_guru_r2r’:
      pyfftw/pyfftw.c:13508:175: error: passing argument 7 of ‘fftwl_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13508 |   __pyx_r = ((void *)fftwl_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((long double *)__pyx_v__in), ((long double *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                               ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                               |
            |                                                                                                                                                                               int *
      /usr/include/fftw3.h:467:1: note: expected ‘const fftwl_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        467 | FFTW_DEFINE_API(FFTW_MANGLE_LONG_DOUBLE, long double, fftwl_complex)
            | ^~~~~~~~~~~~~~~
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyFFTW
Failed to build pyFFTW
ERROR: Could not build wheels for pyFFTW, which is required to install pyproject.toml-based projects
```

GCC 8.5.0 (Rocky 8.9) actually issues the same warning—`int*` and `fftw_r2r_kind*` etc are incompatible—but exits with zero.

Solution would be:

* pass `-Wno-incompatible-pointer-types`
* fix the signatures of `fftwl_plan_guru_r2r`, `fftw_plan_guru_r2r` and  `fftwl_plan_guru_r2r`

### PR DESCRIPTION
**Pull request by e-kwsm**
fix #378

### CODE REVIEW

File: pyfftw/pyfftw.pxd [hunk 1]
```diff
@@ -373,6 +373,48 @@ cdef extern from 'fftw3.h':
 
     double FFTW_NO_TIMELIMIT
 
+    ctypedef enum fftw_r2r_kind:
```
Comment (paugier) [inline pyfftw/pyfftw.pxd:376]: Three times exactly the same `enum`. Why? Is it really needed? Couldn't we use everywhere `fftw_r2r_kind`?

Comment (e-kwsm) [reply pyfftw/pyfftw.pxd:376]: I am not familiar with cython; I tried aliases `fftwf_r2r_kind = fftw_r2r_kind`, `ctypedef fftw_r2r_kind fftwf_r2r_kind` etc, but they do not work, so I naively copy the enum.

I am not sure if `fftw_r2r_kind` can be safely used for `float` and `long double`.

**Comment by paugier** [comment]
@e-kwsm same things here: can you please rebase on master so that the CI is rerun with the versions officially supported (3.10 to 3.13) ?

File: pyfftw/pyfftw.pxd (hunk 1 above)
Comment (paugier) [reply pyfftw/pyfftw.pxd:376]: Alright, let's go like that. At least, it's simple and it works.

**Review by paugier — APPROVED** [review]

**Pull request merged**